chainstate/tx-verifier/src/transaction_verifier/utxos_undo_cache/cached_utxo_block_undo.rs

```diff
@@ -0,0 +1,238 @@
+    }
+
+    pub fn insert_tx_undo(
+        &mut self,
+        tx_id: Id<Transaction>,
+        tx_undo: UtxosTxUndoWithSources,
+    ) -> Result<(), UtxosBlockUndoError> {
+        let (utxos, sources) = tx_undo.consume();
+
+        match self.tx_undos.entry(tx_id) {
+            Entry::Vacant(e) => e.insert(CachedOperation::Write(utxos)),
+            Entry::Occupied(_) => return Err(UtxosBlockUndoError::UndoAlreadyExists(tx_id)),
+        };
+
+        sources
+            .into_iter()
+            .filter_map(|source_id| match source_id {
+                OutPointSourceId::Transaction(id) => Some(id),
+                OutPointSourceId::BlockReward(_) => None,
+            })
+            .for_each(|source_tx_id| {
+                self.child_parent_dependencies
+                    .insert((tx_id, source_tx_id), CachedOperation::Write(()));
+                self.parent_child_dependencies
+                    .insert((source_tx_id, tx_id), CachedOperation::Write(()));
+            });
+
+        Ok(())
+    }
+
+    pub fn has_children_of(&self, tx_id: &Id<Transaction>) -> bool {
+        // Check if the tx is a dependency for other txs.
+        self.parent_child_dependencies
+            .iter()
+            .filter(|((id, _), _)| id == tx_id)
+            .any(|(_, op)| op.get().is_some())
+    }
+
+    fn get_parents_of(&self, tx_id: &Id<Transaction>) -> Vec<(Id<Transaction>, Id<Transaction>)> {
+        self.child_parent_dependencies
+            .iter()
+            .filter(|((child, _), _)| child == tx_id)
+            .filter_map(|((child, parent), op)| op.get().map(|_| (*child, *parent)))
+            .collect()
+    }
+
+    pub fn take_block_reward_undo(&mut self) -> Option<UtxosBlockRewardUndo> {
+        self.reward_undo.take().and_then(|op| op.take())
+    }
+
+    pub fn take_tx_undo(
+        &mut self,
+        tx_id: &Id<Transaction>,
+    ) -> Result<Option<UtxosTxUndo>, UtxosBlockUndoError> {
+        if !self.has_children_of(tx_id) {
+            // If not this tx can be taken and returned.
+            // But first, remove itself as a dependency of others.
+            let to_remove = self.get_parents_of(tx_id);
+
+            to_remove.iter().for_each(|(id1, id2)| {
+                self.child_parent_dependencies.insert((*id1, *id2), CachedOperation::Erase);
+                self.parent_child_dependencies.insert((*id2, *id1), CachedOperation::Erase);
+            });
+
+            let res = self.tx_undos.insert(*tx_id, CachedOperation::Erase).and_then(|op| op.take());
+            Ok(res)
+        } else {
+            Err(UtxosBlockUndoError::TxUndoWithDependency(*tx_id))
+        }
+    }
+
+    pub fn set_block_reward_undo(&mut self, reward_undo: UtxosBlockRewardUndo) {
+        debug_assert!(self.reward_undo.is_none());
+        self.reward_undo = Some(CachedOperation::Write(reward_undo));
+    }
+
+    pub fn combine(&mut self, other: Self) -> Result<(), UtxosBlockUndoError> {
```

Comment:
Can you please document the public functions, explaining what they do and their role in the process?